File: valid-parentheses/go/solutions.go
package validparentheses

func isValid(s string) bool {
	sRune := []rune(s)
	match := map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}
	openingBraces := []rune{}
	closingBraces := []rune{}
	openingBracesOrder := []rune{}
	start := sRune[0]
	length := len(sRune)
	end := sRune[len(sRune)-1]

	for key, val := range match {
		closingBraces = append(closingBraces, key)
		openingBraces = append(openingBraces, val)
	}

	if contains(openingBraces, end) || contains(closingBraces, start) {
		return false
	}

	if length%2 != 0 {
		return false
	}

	for _, char := range sRune {
		if contains(openingBraces, char) {
			openingBracesOrder = append(openingBracesOrder, char)
		} else {
			lastOpening := openingBracesOrder[len(openingBracesOrder)-1]

			if match[char] == lastOpening {
				openingBracesOrder = openingBracesOrder[:len(openingBracesOrder)-1]
			} else {
				return false
			}

		}
	}
	if len(openingBracesOrder) == 0 {
		return true
	}

	return false
}

func contains(arr []rune, str rune) bool {
	for _, a := range arr {
		if a == str {
			return true
		}
	}
	return false
}

File: valid-parentheses/js/solutions.js
/**
 * @param {string} s
 * @return {boolean}
 */
 var isValid = function(s) {
  const openingBraces = new Set(['(', '[', '{']);
  const closingBraces = new Set([')', ']', '}']);
  const match = {
    ')': '(',
    '}': '{',
    ']': '['
  }
  
  const openingBracesOrder = [];
  
  if (closingBraces.has(s[0]) || openingBraces.has(s[s.length -1])) {
      // Starts with a closing bracket or ends with an opening bracket
      // so it's invalid
      return false;
  }
  
  if (s.length % 2 !== 0) {
    // Doesn't contain even number of characters so its invalid
    return false;
  }
  
  for (let i = 0; i < s.length; i++) {
    const char = s[i];
    
    if (openingBraces.has(char)) {
      openingBracesOrder.push(char);
    } else {
      const lastOpening = openingBracesOrder[openingBracesOrder.length - 1];
      if (match[char] === lastOpening) {
        openingBracesOrder.pop()
      } else {
        return false
      }
    }
  }
  
  if (openingBracesOrder.length === 0) {
    return true;
  }
  return false;
};
